lang/java/compiler/src/main/java/org/apache/avro/compiler/idl/DocCommentHelper.java

```diff
@@ -0,0 +1,155 @@
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package org.apache.avro.compiler.idl;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+/**
+ * Utility class with {@code ThreadLocal} fields that allow the generated
+ * classes {@link Idl} and {@link IdlTokenManager} to exchange documentation
+ * comments without forcing explicit parsing of documentation comments.
+ *
+ * The reason this works is that all calls to this class happen within a call to
+ * the method {@link Idl#CompilationUnit()} (either directly or indirectly).
+ */
+public class DocCommentHelper {
+  /**
+   * Pattern to match the common whitespace indents in a multi-line String.
+   * Doesn't match a single-line String, fully matches any multi-line String.
+   *
+   * To use: match on a {@link String#trim() trimmed} String, and then replace all
+   * newlines followed by the group "indent" with a newline.
+   */
+  private static final Pattern WS_INDENT = Pattern.compile("(?U).*\\R(?<indent>\\h*).*(?:\\R\\k<indent>.*)*");
+  /**
+   * Pattern to match the whitespace indents plus common stars (1 or 2) in a
+   * multi-line String. If a String fully matches, replace all occurrences of a
+   * newline followed by whitespace and then the group "stars" with a newline.
+   *
+   * Note: partial matches are invalid.
+   */
+  private static final Pattern STAR_INDENT = Pattern.compile("(?U)(?<stars>\\*{1,2}).*(?:\\R\\h*\\k<stars>.*)*");
+
+  private static final ThreadLocal<DocComment> DOC = new ThreadLocal<>();
+  private static final ThreadLocal<List<String>> WARNINGS = ThreadLocal.withInitial(ArrayList::new);
+
+  /**
+   * Return all warnings that were encountered while parsing, once. Subsequent
+   * calls before parsing again will return an empty list.
+   */
+  static List<String> getAndClearWarnings() {
+    List<String> warnings = WARNINGS.get();
+    WARNINGS.remove();
+    return warnings;
+  }
+
+  static void setDoc(Token token) {
+    DocComment newDocComment = new DocComment(token);
+    DocComment oldDocComment = DOC.get();
+    if (oldDocComment != null) {
+      WARNINGS.get().add(String.format(
+          "Found documentation comment at line %d, column %d. Ignoring previous one at line %d, column %d: \"%s\"\n"
+              + "A common cause is to use documentation comments ( /** ... */ ) instead of multiline comments ( /* ... */ ).",
```

Comment:
Here and below, I think it's more grammatical to say `A common cause is using documentation comments...`

It might even be clearer to say `Did you mean to use a multiline comment ( /* ... */` ) instead?`

What do you think?

Reply:
I prefer the latter: it's nicely concise.